src/click/shell_completion.py

```diff
@@ -0,0 +1,472 @@
+
+    for type key descr in ${response}; do
+      if [[ "$type" == "plain" ]]; then
+        if [[ "$descr" == "_" ]]; then
+          completions+=("$key")
+        else
+          completions_with_descriptions+=("$key":"$descr")
+        fi
+      elif [[ "$type" == "dir" ]]; then
+        _path_files -/
+      elif [[ "$type" == "file" ]]; then
+        _path_files -f
+      fi
+    done
+
+    if [ -n "$completions_with_descriptions" ]; then
+        _describe -V unsorted completions_with_descriptions -U
+    fi
+
+    if [ -n "$completions" ]; then
+        compadd -U -V unsorted -a completions
+    fi
+    compstate[insert]="menucomplete"
+}
+
+compdef %(complete_func)s %(script_names)s
+"""
+
+
+COMPLETION_SCRIPT_FISH = """
+function %(complete_func)s_complete;
+    set -l response;
+    for value in (env %(autocomplete_var)s=complete_fish \
+        COMP_WORDS=(commandline -cp) COMP_CWORD=(commandline -t) \
+        %(script_names)s);
+        set response $response $value;
+    end;
+
+    for completion in $response;
+        set -l metadata (string split "," $completion);
+        if test $metadata[1] = "dir";
+            __fish_complete_directories $metadata[2];
+        else if test $metadata[1] = "file";
+            __fish_complete_path $metadata[2];
+        else if test $metadata[1] = "plain";
+            echo $metadata[2];
+        end;
+    end;
+end;
+
+complete --no-files --command %(script_names)s --arguments \
+    "(%(complete_func)s_complete)"
+"""
+
+_invalid_ident_char_re = re.compile(r"[^a-zA-Z0-9_]")
+
+
+def start_of_option(param_str):
+    """Returns whether or not this si the start of an option declaration
+    (i.e, starts with "-" or "--")
+
+    :param param_str: param_str to check
+    """
+    return param_str and param_str[:1] == "-"
+
+
+class ShellComplete(ABC):
+    """The ShellComplete class acts as the minimal API contract of a
+    completion class.  It is not intended to be used directly as
+    :meth:`source` and :meth:`complete` must be overridden.
+
+
+    :param prog_name: the program name that should be used.  By default
+                        the program name is constructed by taking the
+                        file name from ``sys.argv[0]``.
+
+    :param complete_var: the environment variable used to activate
+                         shell completion.
+    :param cli: command definition
+    """
+
+    def __init__(self, prog_name, complete_var, cli):
+        # set shell-specific environment variables
+        self.prog_name = prog_name
+        self.complete_var = complete_var
+        self.cli = cli
+
+    @abstractmethod
+    def source(self):
+        """Returns the string to be echoed during activation, and is
+        automatically invoked by :func:`shell_complete`. The string
+        will also be formatted with the variables ``script_name``,
+        ``autocomplete_var`` and ``complete_func`` first.
+
+        The default implementation is raising a not implemented error.
```

Comment:
```suggestion
        The default implementation is raising a `NotImplemented` error.
```

Reply:
Removed this line completely as it's outdated. Thanks. (same goes for https://github.com/pallets/click/pull/1622#discussion_r465627296)